[addons][tools] add thread helper classes to addon headers

AlwinEsch opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->
Next step to improve dev kit headers and to make easier to remove addon own parts or depends (e.g. p8-platform)

Commit 1:
----------------------
**[addons][tools] add helper header for thread processing**

This adds another help header to easily create threads on addons and is based entirely on the thread support that has existed since C++11, therefore OS independent.

It is a further step to remove the p8-platform usage of addons. As a result of the fact that Kodi and addons C++17 is now coming and p8-platform is based on C++, correct use cannot be guaranteed, as it is not certain whether there will also be C++17.

Code is also partly taken from Kodi.

Commit 2:
----------------------
**[addons][tools] add documentation to new CThread addon helper class**

This adds new documentation to the thread class so that it can be seen
by the user and processed using doxygen.

Commit 3:
----------------------
**[addons][tools] add timer thread class**

This adds a class to the add-on to introduce functions for the time interval call.

This is also based on the original but revised to use C++ correctly.

Commit 4:
----------------------
**[addons][tools] add documentation to new CTimer class**

This add the required documentation to have usable on doxygen and give for addon devs.

Commit 5:
----------------------
**[addons][tools] increase tools API version to 1.0.2**

Related to commits before.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][tools] add documentation to new CThread addon helper class

This adds new documentation to the thread class so that it can be seen
by the user and processed using doxygen.

AlwinEsch committed:
[addons][tools] increase tools API version to 1.0.2

Related to commits before.

garbear commented on xbmc/addons/kodi-dev-kit/include/kodi/tools/Timer.h:
```diff
@@ -0,0 +1,290 @@
+class CTimer : protected CThread
+{
+public:
+  class ITimerCallback
```
Missing semicolon?

AlwinEsch replied:
Ah yes, funny that has compile has worked at home :thinking:

AlwinEsch committed:
[addons][tools] add timer thread class

This adds a class to the add-on to introduce functions for the time interval call.

This is also based on the original but revised to use C++ correctly.

AlwinEsch committed:
[addons][tools] add documentation to new CTimer class

This add the required documentation to have usable on doxygen and
give for addon devs.

garbear reviewed (approved):
I didn't check the threading and timer logic in-depth, but it looks like the code in core that I'm familiar with. I'll assume it has sufficiently few edge-case errors for add-ons.

ksooo commented on xbmc/addons/kodi-dev-kit/include/kodi/tools/Thread.h:
```diff
@@ -0,0 +1,398 @@
+/// ~~~~~~~~~~~~~
+///
+///@{
+class CThread
```
May I suggest to name the class just `Thread`, without the `C`prefix? As far as I know we do not use C prefix for devkit classes.

AlwinEsch replied:
In the main classes I use "C..." only when it came to additional auxiliary data I had left out C prefix (e.g. with PVRChannel or AudioEngineFormat) because there only values are changed but no processes are (so to speak, extended structure).

ksooo replied:
okay. your decision.

ksooo commented on xbmc/addons/kodi-dev-kit/include/kodi/tools/Thread.h:
```diff
@@ -0,0 +1,398 @@
+    {
+      // it's possible that the thread exited on it's own without a call to StopThread. If so then
+      // the promise should be fulfilled.
+      std::future_status stat = m_future.wait_for(std::chrono::milliseconds(0));
```
`#include <chrono>`is missing.

AlwinEsch replied:
Thanks a lot for see. Have added the missing header.

ksooo commented on xbmc/addons/kodi-dev-kit/include/kodi/tools/Timer.h:
```diff
@@ -0,0 +1,315 @@
+/// ~~~~~~~~~~~~~
+///
+///@{
+class CTimer : protected CThread
```
Just `Timer`, not `CTimer`?

phunkyfish replied:
Just `Timer` will cause conflicts with some PVR add-one I would say.

AlwinEsch committed:
[addons][tools] add helper header for thread processing

This adds another help header to easily create threads on addons and is based entirely
on the thread support that has existed since C++11, therefore OS independent.

It is a further step to remove the p8-platform usage of addons. As a result of the fact
that Kodi and addons C++17 is now coming and p8-platform is based on C++, correct use
cannot be guaranteed, as it is not certain whether there will also be C++17.

Code is also partly taken from Kodi.

phunkyfish reviewed (approved):
👍
